demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,613 @@
+            }
+
+            if( xClass != CKO_PUBLIC_KEY )
+            {
+                configPRINTF( ( "Incorrect object class returned from GetAttributeValue.\r\n" ) );
+            }
+
+            /* Check that both keys are stored as EC Keys. */
+            xTemplate.type = CKA_KEY_TYPE;
+            xTemplate.pValue = &xKeyType;
+            xTemplate.ulValueLen = sizeof( CK_KEY_TYPE );
+            xResult = pxGlobalFunctionList->C_GetAttributeValue( xGlobalSession, xPrivateKeyHandle, &xTemplate, 1 );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Error getting attribute value of EC key type.\r\n" ) );
+            }
+
+            if( xTemplate.ulValueLen != sizeof( CK_KEY_TYPE ) )
+            {
+                configPRINTF( ( "Length of key type incorrect in GetAttributeValue.\r\n" ) );
+            }
+
+            if( xKeyType != CKK_EC )
+            {
+                configPRINTF( ( "Incorrect key type for private key.\r\n" ) );
+            }
+
+            xResult = pxGlobalFunctionList->C_GetAttributeValue( xGlobalSession, xPublicKeyHandle, &xTemplate, 1 );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Error getting attribute value of EC key type.\r\n" ) );
+            }
+
+            if( xTemplate.ulValueLen != sizeof( CK_KEY_TYPE ) )
+            {
+                configPRINTF( ( "Length of key type incorrect in GetAttributeValue.\r\n" ) );
+            }
+
+            if( xKeyType != CKK_EC )
+            {
+                configPRINTF( ( "Incorrect key type for public key.\r\n" ) );
+            }
+
+            /* Check that public key point can be retrieved for public key. */
+            xTemplate.type = CKA_EC_POINT;
+            xTemplate.pValue = xEcPoint;
+            xTemplate.ulValueLen = sizeof( xEcPoint );
+            xResult = pxGlobalFunctionList->C_GetAttributeValue( xGlobalSession, xPublicKeyHandle, &xTemplate, 1 );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to retrieve EC Point.\r\n" ) );
+            }
+
+            /* Perform a sign with the generated private key. */
+            xMechanism.mechanism = CKM_ECDSA;
+            xMechanism.pParameter = NULL;
+            xMechanism.ulParameterLen = 0;
+            xResult = pxGlobalFunctionList->C_SignInit( xGlobalSession, &xMechanism, xPrivateKeyHandle );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to SignInit ECDSA.\r\n" ) );
+            }
+
+            xSignatureLength = sizeof( xSignature );
+            xResult = pxGlobalFunctionList->C_Sign( xGlobalSession, xHashedMessage, SHA256_DIGEST_SIZE, xSignature, &xSignatureLength );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to ECDSA Sign.\r\n" ) );
+            }
+
+            /* Verify the signature with mbedTLS */
```

Comment:
I think that round-tripping the new key between a PKCS #11 signature and an mbedTLS verification would in fact be a great addition here. It doesn't look like that's happening, though, at least not in this immediate few lines of code.

Reply:
Added!